teps/0076-array-result-types.md

```diff
@@ -0,0 +1,321 @@
+  - name: get-environments
+  - name: deploy
+    params:
+      - name: environment
+        value: '$(tasks.get-environments.environments[0])'
+```
+
+## Test Plan
+
+In addition to unit tests the implementation would include:
+
+* At least one reconciler level integration test which includes an array result
+* A tested example of a Task that produces an array result with:
+  * Another Task consuming that array
+  * Another Task consuming one specific value from that array
+
+## Design Evaluation
+
+* [Reusability](https://github.com/tektoncd/community/blob/main/design-principles.md#reusability):
+  * Pro: This will improve the reusability of Tekton components by enabling the scenario of a Task providing an array result
+    and then a Pipeline being able to loop over those value for subsequent Tasks (i.e. reusing Tasks which are designed
+    for single inputs, but users want to use 'for each' item in an array)
+* [Simplicity](https://github.com/tektoncd/community/blob/main/design-principles.md#simplicity)
+  * Pro: This proposal reuses the existing array or string concept for params
+  * Pro: This proposal continues [the precedent of using JSONPath syntax in variable replacement](https://github.com/tektoncd/pipeline/issues/1393#issuecomment-561476075)
+  * Con: There are now two different files/paths to which users can write results and they have to understand when
+    to use which. (We could choose to deprecate the original path if necessary.)
+* [Flexibility](https://github.com/tektoncd/community/blob/main/design-principles.md#flexibility)
+  * Con: Although there is a precedent for including JSONPath syntax, this is a step toward including more hard coded
+    expression syntax in the Pipelines API (without the ability to choose other language options)
+* [Conformance](https://github.com/tektoncd/community/blob/main/design-principles.md#conformance)
+  * Supporting array results and indexing syntax would be included in the conformance surface
+
+## Drawbacks
+
+* Increases the size of TaskRuns (by including arrays in termination messages and results)
+* Only partial support for JSONPath (i.e. just array indexing support, nothing more)
+
+## Alternatives
+
+1. Do not support arrays explicitly - if we do this, we probably shouldn't do TEP-0075 either, i.e. results should
+  always just be strings
+1. Add support for array results, but don't add support for indexing into arrays
+   * Con: In TEP-0075 we propose adding support for dictionaries; if we take the same approach as this proposal, this
+     would mean not allowing the ability to get individual values out of those dictionaries which would significantly
+     limit the number of Tasks this would be compatible with (i.e. we should be consistent between arrays and
+     dictionaries)
+   * Con: Without indexing, arrays will only work with arrays. This means you cannot combine arrays with Tasks that
+     only take string parameters and this limits the interoperabiltiy between Tasks.
```

Comment:
It makes sense, I kind of struggle to imaging an actual example though.
I would need an array parameter where certain positions in the array hold special meaning I guess?